### ISSUE TITLE
CMake module: false behaviour of CMakeTraceParser._guess_files()

### ISSUE THREAD

**Issue by gerion0** [mesonbuild/meson#7601]
**Describe the bug**
Meson fails to build a CMake project with `add_custom_command`.
I'll start with a minimal example, then the description is much easier.

**To Reproduce**

Execute this steps:
```
$ mkdir mwe && cd mwe
$ mkdir subprojects
$ git clone https://github.com/metabrainz/libmusicbrainz subprojects/libmusicbrainz
```
Then create the following meson.build file:
```
project('mwe', 'cpp',
  version : '0.1',
  default_options : ['warning_level=3',
                     'cpp_std=c++17'])

cmake = import('cmake')
libmusicbrainz_proj = cmake.subproject('libmusicbrainz')
libmusicbrainz = libmusicbrainz_proj.dependency('musicbrainz5')
```
Building this project fails (I'm using meson git head):
```
% ~/sourcecode/meson/meson.py build
The Meson build system
Version: 0.55.999
Source dir: /home/ci3nt/test/mwe
Build dir: /home/ci3nt/test/mwe/build
Build type: native build
Project name: mwe
Project version: 0.1
C++ compiler for the host machine: ccache c++ (gcc 9.3.0 "c++ (Gentoo 9.3.0 p2) 9.3.0")
C++ linker for the host machine: c++ ld.bfd 2.33.1
Host machine cpu family: x86_64
Host machine cpu: x86_64

|Executing subproject libmusicbrainz method cmake 
|
|Found CMake: /usr/bin/cmake (3.16.5)
|
||Configuring the build directory with CMake version 3.16.5
||Running: -DCMAKE_CXX_COMPILER_LAUNCHER=ccache -DCMAKE_CXX_COMPILER=c++ -G Ninja -DCMAKE_INSTALL_PREFIX=/usr/local
||  - build directory:          /home/ci3nt/test/mwe/build/subprojects/libmusicbrainz/__CMake_build
||  - source directory:         /home/ci3nt/test/mwe/subprojects/libmusicbrainz
||  - trace args:               --trace --trace-expand --no-warn-unused-cli --trace-redirect=cmake_trace.txt
||  - preload file:             /home/ci3nt/test/mwe/build/meson-private/data/preload.cmake
||  - disabled policy warnings: [CMP0025, CMP0047, CMP0056, CMP0060, CMP0065, CMP0066, CMP0067, CMP0082, CMP0089]
||
||Running with trace output on.
||Running with expanded trace output on.
||Not searching for unused variables given on the command line.
||Trace will be written to cmake_trace.txt
||-- The C compiler identification is GNU 9.3.0
||-- The CXX compiler identification is GNU 9.3.0
||-- Check for working C compiler: /usr/bin/cc
||-- Check for working C compiler: /usr/bin/cc -- works
||-- Detecting C compiler ABI info
||-- Detecting C compiler ABI info - done
||-- Detecting C compile features
||-- Detecting C compile features - done
||-- Check for working CXX compiler: /usr/bin/c++
||-- Check for working CXX compiler: /usr/bin/c++ -- works
||-- Detecting CXX compiler ABI info
||-- Detecting CXX compiler ABI info - done
||-- Detecting CXX compile features
||-- Detecting CXX compile features - done
||-- Found PkgConfig: /usr/bin/pkg-config (found version "0.29.2") 
||-- Checking for module 'neon'
||--   Found neon, version 0.30.2
||-- Found Neon: /usr/lib/libneon.so  
||-- Checking for module 'libxml-2.0'
||--   Found libxml-2.0, version 2.9.9
||-- Found LibXml2: /usr/lib/libxml2.so  
||CMake Warning (dev) at /home/ci3nt/test/mwe/build/meson-private/data/preload.cmake:46 (if):
||  Policy CMP0054 is not set: Only interpret if() arguments as variables or
||  keywords when unquoted.  Run "cmake --help-policy CMP0054" for policy
||  details.  Use the cmake_policy command to set the policy and suppress this
||  warning.
||
||  Quoted variables like "PROPERTIES" will no longer be dereferenced when the
||  policy is set to NEW.  Since the policy is not set the OLD behavior will be
||  used.
||Call Stack (most recent call first):
||  tests/CMakeLists.txt:23 (set_source_files_properties)
||This warning is for project developers.  Use -Wno-dev to suppress it.
||
||-- Configuring done
||-- Generating done
||-- Build files have been written to: /home/ci3nt/test/mwe/build/subprojects/libmusicbrainz/__CMake_build
|
|CMake configuration: SUCCEEDED
|CMake project libmusicbrainz5 has 14 build targets.
|
||Processing generated meson AST
||Build file: /home/ci3nt/test/mwe/build/subprojects/libmusicbrainz/meson.build
||
||WARNING: In subproject libmusicbrainz: Unknown options: "libmusicbrainz:cpp_std"
||The value of new options can be set with:
||meson setup <builddir> --reconfigure -Dnew_option=new_value ...
||Project name: libmusicbrainz5
||Project version: undefined
||C compiler for the host machine: ccache cc (gcc 9.3.0 "cc (Gentoo 9.3.0 p2) 9.3.0")
||C linker for the host machine: cc ld.bfd 2.33.1
||C++ compiler for the host machine: ccache c++ (gcc 9.3.0 "c++ (Gentoo 9.3.0 p2) 9.3.0")
||C++ linker for the host machine: c++ ld.bfd 2.33.1
||Build targets in project: 14
||Subproject libmusicbrainz finished.
|

Build targets in project: 14

mwe 0.1

  Subprojects
    libmusicbrainz: YES

Found ninja-1.10.0 at /usr/bin/ninja
% cd build && ninja
[27/52] Generating cm_custom_tgt_0 with a custom command
FAILED: subprojects/libmusicbrainz/mb5_c.cc subprojects/libmusicbrainz/mb5_c.h subprojects/libmusicbrainz/c0_mb5_c.h 
/usr/lib/python-exec/python3.8/python3 /home/ci3nt/sourcecode/meson/meson.py --internal cmake_run_ctgt -o subprojects/libmusicbrainz/mb5_c.cc subprojects/libmusicbrainz/mb5_c.h subprojects/libmusicbrainz/c0_mb5_c.h -O /home/ci3nt/test/mwe/build/subprojects/libmusicbrainz/__CMake_build/src/mb5_c.cc /home/ci3nt/test/mwe/build/subprojects/libmusicbrainz/__CMake_build/src/mb5_c.h /home/ci3nt/test/mwe/build/subprojects/libmusicbrainz/__CMake_build/src/../include/musicbrainz5/mb5_c.h -d /home/ci3nt/test/mwe/build/subprojects/libmusicbrainz/__CMake_build/src /home/ci3nt/test/mwe/build/subprojects/libmusicbrainz/cm_make_c_interface '/home/ci3nt/test/mwe/subprojects/libmusicbrainz/src cinterface.xml' '/home/ci3nt/test/mwe/build/subprojects/libmusicbrainz/__CMake_build/src mb5_c.cc' mb5_c.h ';;;' /usr/bin/cmake -E copy_if_different /home/ci3nt/test/mwe/build/subprojects/libmusicbrainz/__CMake_build/src/mb5_c.h /home/ci3nt/test/mwe/build/subprojects/libmusicbrainz/__CMake_build/src/../include/musicbrainz5/mb5_c.h ';;;'
/home/ci3nt/test/mwe/build/subprojects/libmusicbrainz/cm_make_c_interface /home/ci3nt/test/mwe/subprojects/libmusicbrainz/src cinterface.xml /home/ci3nt/test/mwe/build/subprojects/libmusicbrainz/__CMake_build/src mb5_c.cc mb5_c.h 
Error copying file (if different) from "/home/ci3nt/test/mwe/build/subprojects/libmusicbrainz/__CMake_build/src/mb5_c.h" to "/home/ci3nt/test/mwe/build/subprojects/libmusicbrainz/__CMake_build/src/../include/musicbrainz5/mb5_c.h".
[44/52] Compiling C++ object subprojects/libmusicbrainz/libcm_musicbrainz5cc.so.p/src_ISWCList.cc.o
ninja: build stopped: subcommand failed.
```

**Expected behavior**
The build should succeed.

**System parameters**
* Is this a [cross build](https://mesonbuild.com/Cross-compilation.html) or just a plain native build (for the same computer)?
Native build

* what operating system (e.g. MacOS Catalina, Windows 10, CentOS 8.0, Ubuntu 18.04, etc.)
Gentoo

* what Python version are you using e.g. 3.8.0
3.8.4

* what `meson --version`
Git HEAD
* what `ninja --version` if it's a Ninja build
1.10.0

**Additional information**
I tried to find out why the build fails. The point is that the meson wrapper program (cmake_run_ctgt) tries to execute:
```
mwe/build/subprojects/libmusicbrainz/cm_make_c_interface 'mwe/subprojects/libmusicbrainz/src cinterface.xml' 'mwe/build/subprojects/libmusicbrainz/__CMake_build/src mb5_c.cc' mb5_c.h
```
This command fails. It should generate the files `mb5_c.cc` and `mb5_c.h`, but does not. Since this command fails silently the `cmake` command after that (`/usr/bin/cmake -E copy_if_different ...`) fails.

The appropriate part in the `CMakeLists.txt` is:
```
ADD_CUSTOM_COMMAND(
	OUTPUT ${CMAKE_CURRENT_BINARY_DIR}/mb5_c.cc ${CMAKE_CURRENT_BINARY_DIR}/mb5_c.h ${CMAKE_CURRENT_BINARY_DIR}/../include/musicbrainz5/mb5_c.h
	COMMAND make-c-interface ${CMAKE_CURRENT_SOURCE_DIR} cinterface.xml ${CMAKE_CURRENT_BINARY_DIR} mb5_c.cc mb5_c.h
	COMMAND ${CMAKE_COMMAND} -E copy_if_different 	${CMAKE_CURRENT_BINARY_DIR}/mb5_c.h ${CMAKE_CURRENT_BINARY_DIR}/../include/musicbrainz5/mb5_c.h
	DEPENDS make-c-interface cinterface.xml c-int-medium-defines.inc c-int-query-source.inc c-int-source-funcs.inc
                c-int-medium-source.inc c-int-release-defines.inc c-int-query-defines.inc c-int-release-source.inc
)
```
Here you can also see the main problem. `make-c-interface` expects 6 arguments that it gets in CMake but not in Meson. In Meson 4 arguments are magically merged: `'mwe/subprojects/libmusicbrainz/src cinterface.xml'` and `'mwe/build/subprojects/libmusicbrainz/__CMake_build/src mb5_c.cc'`. I traced this back to the function [`CMakeParser._guess_files()`](https://github.com/mesonbuild/meson/blob/master/mesonbuild/cmake/traceparser.py#L653). I guess, this function tries to merge paths and filenames. This is probably necessary in other cases, but causes the error here. Is this fixable? Is is otherwise possible to introduce a flag that deactivates this behaviour?

**Comment by mensinda** [comment]
First of all, thank you for the detailed bug report. Yes, `_guess_files()` is indeed the problem in this case. This function is a best-effort for handling whitespaces in the pre-CMake 3.17 trace format. Upgrading to CMake 3.17 should solve this (and is also recommended in general) since meson can then use the new JSON trace format.

A bit of background information:

CMake <3.17 outputs the following trace line for the `ADD_CUSTOM_COMMAND` (see the `cmake_trace.txt` in the `build/subprojects/*/__CMake_build` directory):

```
/home/daniel/mwe/subprojects/libmusicbrainz/src/CMakeLists.txt(36):  ADD_CUSTOM_COMMAND(OUTPUT /home/daniel/mwe/build/subprojects/libmusicbrainz/__CMake_build/src/mb5_c.cc /home/daniel/mwe/build/subprojects/libmusicbrainz/__CMake_build/src/mb5_c.h /home/daniel/mwe/build/subprojects/libmusicbrainz/__CMake_build/src/../include/musicbrainz5/mb5_c.h COMMAND make-c-interface /home/daniel/mwe/subprojects/libmusicbrainz/src cinterface.xml /home/daniel/mwe/build/subprojects/libmusicbrainz/__CMake_build/src mb5_c.cc mb5_c.h COMMAND /home/daniel/prefix/cmake/bin/cmake -E copy_if_different /home/daniel/mwe/build/subprojects/libmusicbrainz/__CMake_build/src/mb5_c.h /home/daniel/mwe/build/subprojects/libmusicbrainz/__CMake_build/src/../include/musicbrainz5/mb5_c.h DEPENDS make-c-interface cinterface.xml c-int-medium-defines.inc c-int-query-source.inc c-int-source-funcs.inc c-int-medium-source.inc c-int-release-defines.inc c-int-query-defines.inc c-int-release-source.inc )
```

Note that there is absolutely no indication when a whitespace char separates arguments or when a whitespace char is part of a path. Hence, the `_guess_files()` function.
Besides this obvious problem, it is also a very lovely syntax to parse in general. I especially like that newlines are not escaped so that there can be linebreaks in a single trace line.

However, this changes with CMake >=3.17 where the trace output is:

```json
{
  "args": [
    "OUTPUT",
    "/home/daniel/mwe/build/subprojects/libmusicbrainz/__CMake_build/src/mb5_c.cc",
    "/home/daniel/mwe/build/subprojects/libmusicbrainz/__CMake_build/src/mb5_c.h",
    "/home/daniel/mwe/build/subprojects/libmusicbrainz/__CMake_build/src/../include/musicbrainz5/mb5_c.h",
    "COMMAND",
    "make-c-interface",
    "/home/daniel/mwe/subprojects/libmusicbrainz/src",
    "cinterface.xml",
    "/home/daniel/mwe/build/subprojects/libmusicbrainz/__CMake_build/src",
    "mb5_c.cc",
    "mb5_c.h",
    "COMMAND",
    "/usr/bin/cmake",
    "-E",
    "copy_if_different",
    "/home/daniel/mwe/build/subprojects/libmusicbrainz/__CMake_build/src/mb5_c.h",
    "/home/daniel/mwe/build/subprojects/libmusicbrainz/__CMake_build/src/../include/musicbrainz5/mb5_c.h",
    "DEPENDS",
    "make-c-interface",
    "cinterface.xml",
    "c-int-medium-defines.inc",
    "c-int-query-source.inc",
    "c-int-source-funcs.inc",
    "c-int-medium-source.inc",
    "c-int-release-defines.inc",
    "c-int-query-defines.inc",
    "c-int-release-source.inc"
  ],
  "cmd": "ADD_CUSTOM_COMMAND",
  "file": "/home/daniel/mwe/subprojects/libmusicbrainz/src/CMakeLists.txt",
  "frame": 1,
  "line": 36,
  "time": 1597565086.734956
}
```

which is an easy to parse JSON document and meson doesn't have to worry about guessing whitespaces. So upgrading to CMake 3.17 (or 3.18) should solve your problem and make the entire trace logic more robust in general.

Regarding your original question:

> Is this fixable?

Maybe with a lot of work. However, I am not sure if this is worth the effort, and changing the logic can lead to possible regressions for other projects. The problem is also that this code can't be made to work 100% perfectly because of the inherent problems with the old trace format. So this function is more of a nice to have that works for most projects prior to CMake 3.17. However, if it breaks a CMake upgrade is definitely the way to go.

> Is is otherwise possible to introduce a flag that deactivates this behaviour?

We generally try our best to minimize the number of configuration options, so the answer will be a technically yes, but practically no. This is also because this problem will fix itself over time when CMake >=3.17 is more widespread. Once this happens we will be stuck with a useless configuration option that we can't remove without breaking the setup of someone who now depends on it...

**Comment by gerion0** [comment]
Thanks for the explanation. When a simple upgrade of CMake helps, there is no need to fix/change anything on the Meson side.

I thing, it may be meaningful to warn the user about that. I have prepared a pull request for that.
